### PR TITLE
Move GetMethodTable and IsReference to JIT

### PR DESCRIPTION
**Pull request by MichalPetryka**
Moves the simple and most important RuntimeHelpers intrinsics to the JIT.

Extracted from #88543.

### CODE REVIEW

**Comment by MichalPetryka** [comment]
@MihuBot

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -4699,6 +4699,17 @@ GenTree* Compiler::optAssertionProp_Update(GenTree* newTree, GenTree* tree, Stat
             if (parent != nullptr)
             {
                 parent->ReplaceOperand(useEdge, newTree);
+                if ((parent->gtOper == GT_IND) && newTree->IsIconHandle())
+                {
+                    parent->gtFlags |= GTF_IND_NONFAULTING;
+                    GenTreeFlags handleKind = newTree->GetIconHandleFlag();
+
+                    if ((handleKind != GTF_ICON_STATIC_HDL) && (handleKind != GTF_ICON_BBC_PTR) &&
```
Comment (EgorBo) [inline src/coreclr/jit/assertionprop.cpp:4707]: this, gtNewIndir and debug-check have these condition - can you unify them all into a helper method or something?

File: src/coreclr/jit/importercalls.cpp [hunk 2]
```diff
@@ -2759,6 +2759,32 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
...
+            }
+
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsReferenceOrContainsReferences:
+            {
+                assert(sig->sigInst.methInstCount == 1);
+                CORINFO_CLASS_HANDLE hClass = sig->sigInst.methInst[0];
+                retNode =
+                    gtNewIconNode(((info.compCompHnd->getClassAttribs(hClass) &
+                                    (CORINFO_FLG_VALUECLASS | CORINFO_FLG_CONTAINS_GC_PTR)) == CORINFO_FLG_VALUECLASS)
+                                      ? 0
```
Comment (EgorBo) [inline src/coreclr/jit/importercalls.cpp:2777]: can you re-format your code to avoid this ugly formatting artifact

Comment (MichalPetryka) [reply src/coreclr/jit/importercalls.cpp:2777]: That's how `jit-format` wants it to look.

File: src/coreclr/jit/assertionprop.cpp (hunk 1 above)
Comment (MichalPetryka) [reply src/coreclr/jit/assertionprop.cpp:4707]: What would be the best place for the helper?

File: src/coreclr/jit/importercalls.cpp [hunk 3]
```diff
@@ -2759,6 +2759,32 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
                 break;
             }
 
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsReference:
+            {
+                assert(sig->sigInst.methInstCount == 1);
+                CORINFO_CLASS_HANDLE hClass = sig->sigInst.methInst[0];
+                retNode                     = gtNewIconNode(eeIsValueClass(hClass) ? 0 : 1);
```
Comment (EgorBo) [inline src/coreclr/jit/importercalls.cpp:2766]: isn't eeIsValueClass `false` for unsafe pointers? (can be used via IL)

File: src/coreclr/jit/importercalls.cpp [hunk 4]
```diff
@@ -2759,6 +2759,32 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
...
+                break;
+            }
+
+            case NI_System_Runtime_CompilerServices_RuntimeHelpers_IsReferenceOrContainsReferences:
+            {
+                assert(sig->sigInst.methInstCount == 1);
+                CORINFO_CLASS_HANDLE hClass = sig->sigInst.methInst[0];
+                retNode =
+                    gtNewIconNode(((info.compCompHnd->getClassAttribs(hClass) &
+                                    (CORINFO_FLG_VALUECLASS | CORINFO_FLG_CONTAINS_GC_PTR)) == CORINFO_FLG_VALUECLASS)
```
Comment (EgorBo) [inline src/coreclr/jit/importercalls.cpp:2776]: same here, it seems like it's not exactly the same check as was on VM side?

File: src/coreclr/jit/importercalls.cpp (hunk 2 above)
Comment (EgorBo) [reply src/coreclr/jit/importercalls.cpp:2777]: I know, I mean just spill the flags into a variable or something like that to change that

File: src/coreclr/jit/assertionprop.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/assertionprop.cpp:4707]: Don't know, perhaps some GenTreeIndir::SetIndirFlags() ?

File: src/coreclr/jit/importercalls.cpp (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/jit/importercalls.cpp:2766]: it seems that CoreCLR and CG2/NAOT have different behavior here? cc @jkotas

File: src/coreclr/jit/assertionprop.cpp [hunk 5]
```diff
@@ -4699,6 +4699,17 @@ GenTree* Compiler::optAssertionProp_Update(GenTree* newTree, GenTree* tree, Stat
             if (parent != nullptr)
             {
                 parent->ReplaceOperand(useEdge, newTree);
+                if ((parent->gtOper == GT_IND) && newTree->IsIconHandle())
```
Comment (AndyAyersMS) [inline src/coreclr/jit/assertionprop.cpp:4702]: What has changed that requires modifying assertion prop?

Comment (MichalPetryka) [reply src/coreclr/jit/assertionprop.cpp:4702]: VN can fold the method table indir to a constant handle which this place then replaces indirs on such method table with. This causes an assert in `fgdiagnostic` to fire complaining about the indir missing the invariant and nonfaulting flags.

Comment (AndyAyersMS) [reply src/coreclr/jit/assertionprop.cpp:4702]: But why couldn't this also happen without your changes? Is VN able to recognize something in your expansions but not the inlined IL intrinsics?

Comment (EgorBo) [reply src/coreclr/jit/assertionprop.cpp:4702]: Yes, so basically we had `IND(IND(knownObj))` and the inner `IND(knonwObj)` was folded by assertprop (with help of VN) to a `CLASS_HANDLE` so now we have `IND(CLASS_HANDLE)` but without proper flags (it's not really a correctness issue, more like annoying requirement from the fgdiagnostic)

Comment (MichalPetryka) [reply src/coreclr/jit/assertionprop.cpp:4702]: > Is VN able to recognize something in your expansions but not the inlined IL intrinsics?

I'd assume it's not able to recognize the field address + subtract magic there.

> But why couldn't this also happen without your changes?

I think this should reproduce with any indir over the method table returned from `gtNewMethodTableLookup`, seems like nothing does that today though.

File: src/coreclr/jit/importercalls.cpp (hunk 3 above)
Comment (jkotas) [reply src/coreclr/jit/importercalls.cpp]: Generic instantiation arg cannot be a pointer, so there is no observable behavior difference. There are probably hundreds of places that make similar simplifying assumption, so I am not particularly worried about it.

**Comment by jkotas** [comment]
crossgen crashing during the build with:
```
  /__w/1/s/src/coreclr/jit/fgdiagnostic.cpp:3290
  Assertion failed '!"Missing flags on tree"' in 'Interop+Sys:.cctor()' during 'Insert GC Polls' (IL size 27; hash 0xb6a186f0; FullOpts)
```

File: src/coreclr/jit/importercalls.cpp (hunk 4 above)
Comment (jkotas) [reply src/coreclr/jit/importercalls.cpp]: dtto

File: src/coreclr/jit/assertionprop.cpp [hunk 6]
```diff
@@ -4699,6 +4699,7 @@ GenTree* Compiler::optAssertionProp_Update(GenTree* newTree, GenTree* tree, Stat
             if (parent != nullptr)
             {
                 parent->ReplaceOperand(useEdge, newTree);
+                parent->gtFlags |= gtGetFlagsForOperand(parent->gtOper, newTree);
```
Comment (SingleAccretion) [inline src/coreclr/jit/assertionprop.cpp:4702]: This flag updating call should be moved to this place in morph:

https://github.com/dotnet/runtime/blob/7980417f4c39c730617059ac983e0bcb4d20f419/src/coreclr/jit/morph.cpp#L9568-L9571

File: src/coreclr/jit/gentree.cpp [hunk 7]
```diff
@@ -4889,6 +4889,32 @@ bool Compiler::gtMarkAddrMode(GenTree* addr, int* pCostEx, int* pCostSz, var_typ
...
+    {
+        flags |= GTF_IND_NONFAULTING;
+
+        GenTreeFlags handleKind = op->GetIconHandleFlag();
+        if ((handleKind != GTF_ICON_STATIC_HDL) && (handleKind != GTF_ICON_BBC_PTR) &&
+            (handleKind != GTF_ICON_GLOBAL_PTR))
+        {
+            flags |= GTF_IND_INVARIANT;
+        }
+    }
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp:4914]: This invariantness setting is fragile; it does not check what type is being loaded.

File: src/coreclr/jit/gentree.cpp [hunk 8]
```diff
@@ -8214,6 +8240,7 @@ GenTreeBlk* Compiler::gtNewBlkIndir(ClassLayout* layout, GenTree* addr, GenTreeF
 //
 GenTreeIndir* Compiler::gtNewIndir(var_types typ, GenTree* addr, GenTreeFlags indirFlags)
 {
+    indirFlags |= gtGetFlagsForOperand(GT_IND, addr);
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp:8243]: Why is this needed here? The caller is expected to pass the right `indirFlags`.

File: src/coreclr/jit/gentree.cpp (hunk 7 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:4914]: Yes, I had the same concern - like why do we expect invariantess if we acess a byte location with long type

File: src/coreclr/jit/gentree.cpp (hunk 8 above)
Comment (EgorBo) [reply src/coreclr/jit/gentree.cpp:8243]: I bet it's not needed it's just how @MichalPetryka was trying to find the case where it's not set

**Comment by JulieLeeMSFT** [comment]
Hi @MichalPetryka, this is a kindly reminder. The .NET 8 RC1 snap is one week away.

**Comment by BruceForstall** [comment]
@MichalPetryka Can you mark this "Draft" until it is ready for review again?
